# Optimized-sgRNA-Design

![](datafiles/CRISPR-Logo.jpg)

This is a set of software tools for analyzing the on-target and off-target effects of sgRNA spacer inputs. This program is heavily based on the following paper:

Brian Mendoza and Cong T. Trinh. 2017. Enhanced Guide-RNA Design And Targeting Analysis For Precise CRISPR Genome Editing Of Single And Consortia Of Industrially Relevant And Non-Model Organisms. DOI: https://doi.org/10.1093/bioinformatics/btx564

This program is still currently under development. Intended functionality is that this program can take in a metagenome consisting of a large number of organisms' genetic data, along with a desired region of interest to be isolated, and output several 20-base-pair sgRNA spacers ranked in terms of optimality.